test/auto_api/states/universal_properties_test.exs

```diff
@@ -44,4 +44,205 @@ defmodule AutoApi.UniversalPropertiesTest do
 
     assert state == %DoorLocksState{positions: locks, timestamp: datetime}
   end
+
+  describe "property timestamp" do
+    test "converts binary value to state" do
+      bin_state =
+        <<0x01, 3::integer-16, 0, 0, 0x02, 0xA4, 9::integer-16, 18, 6, 8, 16, 8, 2, 0, 120, 0x01>>
+
+      datetime = %DateTime{
+        year: 2018,
+        month: 06,
+        day: 08,
+        hour: 16,
+        minute: 8,
+        second: 2,
+        utc_offset: 120,
+        time_zone: "",
+        zone_abbr: "",
+        std_offset: 0
+      }
+
+      state = DiagnosticsState.from_bin(bin_state)
+
+      assert is_map(state.property_timestamps)
+      assert state.property_timestamps[:mileage] == datetime
+    end
+
+    test "converts binary value with multiple items to state" do
+      # NOTE!: Currently elixir doesn't support converting all of the property_timestamps. if an item is repeated multiple times,
+      # we only return the last property's timestamp
+      bin_state =
+        <<4, 0, 2, 2, 1, 4, 0, 2, 3, 0, 164, 0, 11, 18, 11, 28, 12, 56, 13, 0, 0, 4, 2, 1, 164, 0,
+          11, 18, 11, 28, 13, 56, 13, 0, 0, 4, 3, 0>>
+
+      state = DoorLocksState.from_bin(bin_state)
+
+      assert state.positions ==
+               [
+                 %{door_location: :rear_left, position: :closed},
+                 %{door_location: :rear_right, position: :open}
+               ]
+
+      assert state.property_timestamps[:positions] == %DateTime{
+               year: 2018,
+               month: 11,
+               day: 28,
+               hour: 13,
+               minute: 56,
+               second: 13,
+               utc_offset: 0,
+               time_zone: "Etc/UTC",
+               zone_abbr: "UTC",
+               std_offset: 0
+             }
+    end
+
+    test "converts state to binary" do
+      now = DateTime.utc_now()
+
+      state = %DiagnosticsState{
+        mileage: 100,
+        property_timestamps: %{mileage: now},
+        properties: [:mileage, :property_timestamps]
+      }
+
+      assert <<id, _size::integer-16, mileage::integer-24, timestamp_id,
+               timestamp_size::integer-16, timestamp::binary-size(8),
+               id>> = DiagnosticsState.to_bin(state)
+
+      assert id == DiagnosticsState.property_id(:mileage)
+      assert mileage == 100
+      assert timestamp_id == 0xA4
+      assert timestamp_size == 9
+
+      assert timestamp ==
+               <<now.year - 2000, now.month, now.day, now.hour, now.minute, now.second,
+                 now.utc_offset::integer-16>>
+    end
+
+    test "converts state with multiple items to binary" do
+      now = DateTime.utc_now()
+
+      temperature = %{location: :rear_left, temperature: 10.0}
+
+      state = %DiagnosticsState{
+        tire_temperatures: [temperature],
+        property_timestamps: %{tire_temperatures: [{now, temperature}]},
+        properties: [:tire_temperatures, :property_timestamps]
```

Comment:
Do we add `:property_timestamps` here to signal there is something, rather than it being empty?

Reply:
Yes. Since Auto API level 7, vehicles can return only list of properties that they are providing. Since then I introduced `properties` and `to_bin/1` only convert those that are listed in this.

Now that you mentioned, we might be able to change it in the next version and only return the properties that are not nil or empty(list or map)

Reply:
I just looked at line 102 and we do have a rule to not convert empty list or hash to bin. I need to think more about that! I guess in one of the raspberry prototypes that we had we needed to explicitly add properties that we want to parse in `properties`.